File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/preferences/SharedPreferencesBridge.java
package app.zxtune.preferences;

import android.content.SharedPreferences;

import java.util.Map;

import app.zxtune.Releaseable;
import app.zxtune.core.PropertiesModifier;

/**
 * Gate between SharedPreferences and PropertiesModifier
 */
public class SharedPreferencesBridge {

  public static Releaseable subscribe(SharedPreferences preferences, PropertiesModifier target) {
    final RawPropertiesAdapter adapter = new RawPropertiesAdapter(target);
    dumpExistingProperties(preferences.getAll(), adapter);
    SharedPreferences.OnSharedPreferenceChangeListener listener =
        (SharedPreferences updated, String key) -> {
          final Object value = updated.getAll().get(key);
          adapter.setProperty(key, value);
        };
    preferences.registerOnSharedPreferenceChangeListener(listener);
    return () -> preferences.unregisterOnSharedPreferenceChangeListener(listener);
  }

  private static void dumpExistingProperties(Map<String, ?> prefs, RawPropertiesAdapter target) {
    for (Map.Entry<String, ?> entry : prefs.entrySet()) {
      target.setProperty(entry.getKey(), entry.getValue());
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/preferences/SharedPreferencesBridge.kt
package app.zxtune.preferences

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import app.zxtune.Releaseable
import app.zxtune.core.PropertiesModifier

/**
 * Gate between SharedPreferences and PropertiesModifier
 */
object SharedPreferencesBridge {
    @JvmStatic
    fun subscribe(preferences: SharedPreferences, target: PropertiesModifier): Releaseable {
        val adapter = RawPropertiesAdapter(target)
        dumpExistingProperties(preferences.all, adapter)
        val listener =
            OnSharedPreferenceChangeListener { updated: SharedPreferences, key: String? ->
                updated.all[key]?.let { value ->
                    adapter.setProperty(key!!, value)
                }
            }
        preferences.registerOnSharedPreferenceChangeListener(listener)
        return Releaseable { preferences.unregisterOnSharedPreferenceChangeListener(listener) }
    }

    private fun dumpExistingProperties(prefs: Map<String, *>, target: RawPropertiesAdapter) {
        for ((key, value) in prefs) {
            target.setProperty(key, value!!)
        }
    }
}
